src/rest/resource.cpp

```diff
@@ -518,31 +513,22 @@ void Resource::UpdateDiag(std::string aKey, std::vector<otNetworkDiagTlv> &aDiag
 
 void Resource::Diagnostic(const Request &aRequest, Response &aResponse) const
 {
-    otbrError error = OTBR_ERROR_NONE;
+    otThreadSetReceiveDiagnosticGetCallback(mInstance, &Resource::DiagnosticResponseHandler,
+                                            const_cast<Resource *>(this));
     OT_UNUSED_VARIABLE(aRequest);
     struct otIp6Address rloc16address = *otThreadGetRloc(mInstance);
     struct otIp6Address multicastAddress;
 
     VerifyOrExit(otThreadSendDiagnosticGet(mInstance, &rloc16address, kAllTlvTypes, sizeof(kAllTlvTypes)) ==
-                     OT_ERROR_NONE,
-                 error = OTBR_ERROR_REST);
-    VerifyOrExit(otIp6AddressFromString(kMulticastAddrAllRouters, &multicastAddress) == OT_ERROR_NONE,
-                 error = OTBR_ERROR_REST);
+                 OT_ERROR_NONE);
+    VerifyOrExit(otIp6AddressFromString(kMulticastAddrAllRouters, &multicastAddress) == OT_ERROR_NONE);
     VerifyOrExit(otThreadSendDiagnosticGet(mInstance, &multicastAddress, kAllTlvTypes, sizeof(kAllTlvTypes)) ==
-                     OT_ERROR_NONE,
-                 error = OTBR_ERROR_REST);
+                 OT_ERROR_NONE);
 
 exit:
 
-    if (error == OTBR_ERROR_NONE)
-    {
-        aResponse.SetStartTime(steady_clock::now());
-        aResponse.SetCallback();
-    }
-    else
-    {
-        ErrorHandler(aResponse, HttpStatusCode::kStatusInternalServerError);
-    }
+    aResponse.SetStartTime(steady_clock::now());
+    aResponse.SetCallback();
```

Comment:
For now, what happens if we fail in this function? Simply ignore the error?

Reply:
I found that It may fail when node is detached or no buffer for message. 

So I think one solution is to define more HTTP status code according to these newly added error code, I think I could do it after #532 is merged.

Another solution is like this, we ignore all errors(more simple, but still reasonable). 

For one thing, it could address the `no buffer` problem - although we fail in this API call, if we have diagnostic information(received in 4s but due to another call) left in `resource`,  this information is still valid and could be sent as response.
For another thing,  we could just view empty response as  "some issues happen, we can't get anything now".

Reply:
> Simply ignore the error?

Is this true for your changes? I didn't see why we cannot handle the possible failures in this PR. It looks to me that keeping 
```C+++
else
    {
        ErrorHandler(aResponse, HttpStatusCode::kStatusInternalServerError);
    }
```
should just work. Why did you remove this error handler?

Reply:
because if we call `otThreadSendDiagnosticGet` too many times within a specific period, we will get a `no buffer` error for this call( for example, when we send 10 requests for `diagnostics` concurrently , the `no buffer` error usually occurs). 

But actually I prefer not viewing it as an error for HTTP response because the reason for calling `otThreadSendDiagnosticGet` each time when the server received a request for `diagnostics` is to update the information we maintained(we have a hash table to record all the diagnostic information we received within 4 seconds). If we have a 'no buffer' error, it means we have call the API many times recently so information in the hash table is almost the latest.

I know the right way to deal with this problem is to make an exception for `no buffer` rather than remove the error handler.  I expect to do this after #532 because I think I may need to rewrite error handler of all RESTful API according to the modifications in #532 . So here I just ignore it currently.

Reply:
Why is `no buffer error` not good for such situation? I think `no buffer error` was exactly what happened and should be reflected by the HTTP response. 

Maybe we can return `507 Insufficient Storage` for `no buffer error`. 

But if the test is sending too much requests and is expecting `no buffer error` issue to happen sometimes, the test client can conclude with success even if there are some no buffer errors.

Thoughts? @tttttangTH @wgtdkp

Reply:
@tttttangTH You should not just remove the error handler. **You can at least add a log for this error**.

> I know the right way to deal with this problem is to make an exception for no buffer rather than remove the error handler. I expect to do this after #532 because I think I may need to rewrite error handler of all RESTful API according to the modifications in #532 . 

I don't see why we need to wait for #532. Are there any REST feature that depends on #532?

Reply:
Have't found any other feature depend on #532 , but it seems we won't catch `no buffer`  at `otThreadSetReceiveDiagnosticGetCallback` before #532 is applied.

I noticed `no buffer` before when when I wrote this API, at that time, `no buffer` was just an `Info`, and I have chosen to ignore it at that time. But It‘s thoughtless for me to just remove the error handler here.